src/hull.jl

```diff
@@ -1,163 +1,74 @@
-    @assert !isinf(g0.val) "Hull reformulation has failed for non-linear constraint $constr: $gx is not defined at 0. Perspective function is undetermined."
-    g2 = FSG2*g0
-    #create perspective function and simplify
-    pers_func = simplify(g1 - g2, expand = true)
-    #replace FSG expressions & simplify
-    pers_func = substitute(pers_func, Dict(FSG1 => (1-ϵ)*λ+ϵ,
-                                           FSG2 => ϵ*(1-λ)))
-    pers_func = simplify(pers_func)
-    replace_constraint(constr, pers_func, op, rhs)
+function _get_variables(d::Disjunct)
+    union([
+        _get_variables(con)
+        for con in d.constraints
+    ]...)
+end
+function _get_variables(con::JuMP.AbstractArray{<:JuMP.AbstractConstraint})
+    union(_get_variables.(c)...)
+end
+function _get_variables(con::JuMP.ScalarConstraint{JuMP.AffExpr, T}) where {T}
+    collect(keys(con.func.terms))
 end
-hull_reformulation!(constr::AbstractArray{<:ConstraintRef}, bin_var, eps, i, j, k) = 
-    hull_reformulation!(constr[k], bin_var, eps, i, j, k)
 
 """
-    disaggregate_variables(m::Model, disj, bin_var)
 
-Disaggregate all variables in the model and tag them with the disjunction name.
 """
-function disaggregate_variables(m::Model, disj, bin_var)
-    #check that variables are bounded
-    var_refs = m[:gdp_variable_refs]
-    @assert all((has_upper_bound.(var_refs) .&& has_lower_bound.(var_refs)) .|| is_binary.(var_refs)) "All variables must be bounded to perform the Hull reformulation."
-    #reformulate variables
-    obj_dict = object_dictionary(m)
-    bounds_dict = :variable_bounds_dict in keys(obj_dict) ? obj_dict[:variable_bounds_dict] : Dict() #NOTE: should pass as an keyword argument
-    for var_name in m[:gdp_variable_names]
-        var = m[var_name]
-        #define UB and LB
-        LB, UB = get_bounds(var, bounds_dict)
-        #disaggregate variable and add bounding constraints
-        for i in eachindex(disj)
-            var_name_i_str = "$(var_name)_$bin_var$i"
-            var_name_i = Symbol(var_name_i_str)
-            #create disaggregated variable
-            m[var_name_i] = add_disaggregated_variable(m, var, LB, UB, var_name_i_str)
-            #apply bounding constraints on disaggregated variable
-            var_i_lb = "$(var_name_i)_lb" 
-            var_i_ub = "$(var_name_i)_ub" 
-            m[Symbol(var_i_lb)] = @constraint(m, LB * m[bin_var][i] .- m[var_name_i] .<= 0, base_name = var_i_lb)
-            m[Symbol(var_i_ub)] = @constraint(m, m[var_name_i] .- UB * m[bin_var][i] .<= 0, base_name = var_i_ub)
+function _disaggregate_variables(model::JuMP.Model, disj::DisjunctiveConstraintData)
+    disj_vars = _get_variables(disj)
+    for d in disj.constraint.disjuncts
+        #create binary variable for logic variable (indicator)
+        bvar = JuMP.@variable(model, 
+            base_name = string(d.indicator), 
+            binary = true, 
+        )
+        model[Symbol(d.indicator,"_Bin")] = bvar
```

Comment:
Rather than rely on the object dictionary which could have collisions and will make repeated transformations problematic, it would be better to augment `GDPData` to have a mapping dictionary that preserves this information i.e., mapping the disjunct to a variable like this.

This comment also applies to other places in the code that do the same thing, but I am not going to mark all of them.

Reply:
added disaggregated_variables `Dict` to `GDPData`

Reply:
Up to you, think about what would make your life easier with other function calls and functionality.

Reply:
ok. I'll keep them separate for now. merging the two shouldn't be a problem down the road.